eng/pipelines/agent-pool-migration.yml

```diff
@@ -0,0 +1,148 @@
+# The pipeline is mainly used to migrate our agent pool from one version to another version. 
+# Here is how we use the pipeline:
+# 1. What repos are you going to work on for your migration? Put your repos in Parameter 'Repos'. Follow the format in default.
+# 2. The replacement map, e.g We will search all occurrences of the strings in Parameter 'MigrateFrom' and replace to the value in 'MigrateTo'. This is regex map, and you can define multiple replacement rules here.
+# 3. The Parameter 'WhatIfPreference' boolean flag allows you to check your changed files before open a PR. Check the box if you don't want to create a PR.
+# 4. The pipeline will generate PRs in specified repos, Parameter 'ShortForMigrateFrom' and 'ShortForMigrateTo' are the short descriptions of the PR purpose.
+#    The PR title will in the format of "Migrate from ${{parameters.ShortForMigrateFrom}} to ${{parameters.ShortForMigrateTo}}"
+#    You can also specify who triggers the changes in Paramter 'GithubAssigneeAlias'. Then your PR will request review from the specified person.
+#    If you want to create PR from your own fork, you can specify the Parameters 'PRForkOwner'
+# 5. Go to the PR link in 'Create pull request' step. Review and merge the PR. Note, the pipeline intented to be aggressive, so take a closer look at the changes.
+#    The PR also has a link pointed back to the pipeline, so people will learn where it comes from and the purpose of the PR.
+trigger: none
+pr: none
+
+# Notes on ExcludePaths and IncludePaths:
+# 1. ExcludePaths is used to exclude the file pattern you don't want to scan against.
+# 2. IncludePaths is used to add back the files you don't want to exclude. It is supposed to be a subset of ExcludePaths.
+# 3. Both excludePaths and includePaths support standard powershell regex. 
+# 4. The paths here are relative to the repo folder. 
+#    E.g.
+#    In win agent pool, the path is like ".\eng\scripts\Repo-File-Content-Replacements.ps1", so the regex path can be like "^.\\eng\\"
+#    In linux agent pool, the path is like "./eng/scripts/Repo-File-Content-Replacements.ps1", so the regex path can be like "^./eng/"
+#    To compitable with both cases, you can also specify your regex path like "^./|\\eng/|\\"
+parameters:
+- name: Repos
+  type: object
+  default:
+    - RepoOwner: Azure
+      RepoName: azure-sdk-tools
+    - RepoOwner: Azure
+      RepoName: azure-sdk
+      ExcludePaths: '^.\\eng\\common\\*'
```

Comment:
@sima-zhu sorry for drilling on this point, but I do want to stress this is not an expected regex syntax. It works only due to the peculiarities of the implementation, and I don't think this what users would expect.

Looking [here](https://github.com/Azure/azure-sdk-tools/pull/4928/files#diff-292adf4a867d5071fb4e6f56e0895f9c4d35ff88eec4563913ec4f7f1022e7f3R68-R73) we have:

``` ps1
  $relativePath = Resolve-Path -LiteralPath $file.FullName -Relative
  if ($relativePath -notmatch $ExcludePathsRegex) {
    $newFileCollection += $file
  }
  elseif ($relativePath -match $ExcludePathsRegex -and $relativePath -match $IncludePathsRegex) {
    $newFileCollection += $file
  }
```

I ran some tests and here is what I got. Consider following `$relativePath`:

`.\tools\code-owners-parser\ci.yml`

if you would now do, from repo root dir:

`$relativePath -match "^.\\tools\\code-owners-parser\\*"`

it would return true. But only because the `$relativePath` starts with `.`, and that `\*` at the end means "`\`, zero or more times".

Note this also matches (I omitted the `*` at the end).

`$relativePath -match "^.\\tools\\code-owners-parser\\"`

I believe it would be less surprising to the users if it would have proper regex syntax of: `^\\tools\\code-owners-parser\\`, i.e. without the `.` at the beginning and without the `*` at the end. This would match against `$relativePath` of `"\tools\code-owners-parser\ci.yml"` (note the lack of `.`) but not ".\tools\code-owners-parser\ci.yml"`.

Basically the `*` at the end is misused and works only by coincidence, and the usage of `.` at the beginning is nonstandard. `.` denotes "current directory" in the context of file paths, but it denotes "match exactly one character" in the context of regex. Hence possibility of confusion here, as it is used in the context of regex, but to denote the meaning as provided by the path context, not regex context.

Reply:
> Basically the * at the end is misused and works only by coincidence, and the usage of . at the beginning is nonstandard. . denotes "current directory" in the context of file paths, but it denotes "match exactly one character" in the context of regex. Hence what happens here is we use the regex's ., which denotes "any one character", to match the . (i.e. dot character), denoting "current directory", from checked relative path.

I totally agree that we'd better to add escape to match `.` literally and remove the tailing "*". 
However, I have different preference on whether start with `.` or not. To me, start with "." is clearly a relative path, but start with "/" could reference relative or absolute in different context.

Reply:
Sima, sorry if I sounded too insistent - I think just having dot also works - you explained it well in the docs and examples, so I think users will have good experience with that. Next time when I will be commenting I will make sure my preference is communicated better; sorry again! :)